diff --git a/src/eval.py b/src/eval.py
index c82c578..76e5211 100644
--- a/src/eval.py
+++ b/src/eval.py
@@ -33,6 +33,8 @@ def parse_args():
     parser.add_argument("--check_integrity", action="store_true")
     parser.add_argument("--write_out", action="store_true", default=False)
     parser.add_argument("--output_base_path", type=str, default=None)
+    parser.add_argument("--stop_words", type=str, default=None)
+    parser.add_argument("--apply_chat_template", action="store_true", default=False)
 
     return parser.parse_args()
 
@@ -52,6 +54,12 @@ def main():
     else:
         task_names = utils.pattern_match(args.tasks.split(","), tasks.ALL_TASKS)
 
+    if isinstance(args.stop_words, str):
+        if args.stop_words == "":
+            args.stop_words = None
+        else:
+            args.stop_words = args.stop_words.split(",")
+
     print(f"Selected Tasks: {task_names}")
 
     description_dict = {}
@@ -74,7 +82,9 @@ def main():
         check_integrity=args.check_integrity,
         write_out=args.write_out,
         output_base_path=args.output_base_path,
-        model_prompt=args.model_prompt
+        model_prompt=args.model_prompt,
+        stop_words=args.stop_words,
+        apply_chat_template=args.apply_chat_template,
     )
 
     dumped = json.dumps(results, indent=2)
diff --git a/src/evaluator.py b/src/evaluator.py
index f70e801..9cf34c6 100644
--- a/src/evaluator.py
+++ b/src/evaluator.py
@@ -30,7 +30,9 @@ def simple_evaluate(
     decontamination_ngrams_path=None,
     write_out=False,
     output_base_path=None,
-    model_prompt=None
+    model_prompt=None,
+    stop_words=None,
+    apply_chat_template=False,
 ):
     """Instantiate and evaluate a model on a list of tasks.
 
@@ -109,7 +111,9 @@ def simple_evaluate(
         decontamination_ngrams_path=decontamination_ngrams_path,
         write_out=write_out,
         output_base_path=output_base_path,
-        model_prompt=model_prompt
+        model_prompt=model_prompt,
+        stop_words=stop_words,
+        apply_chat_template=apply_chat_template,
     )
 
     # add info about the model and few shot config
@@ -121,6 +125,8 @@ def simple_evaluate(
         "batch_sizes": list(lm.batch_sizes.values()) if hasattr(lm, "batch_sizes") else [],
         "device": device,
         "no_cache": no_cache,
+        "apply_chat_template": apply_chat_template,
+        "stop_words": repr(stop_words),
         "limit": limit,
         "bootstrap_iters": bootstrap_iters,
         "description_dict": description_dict,
@@ -144,7 +150,9 @@ def evaluate(
     decontamination_ngrams_path=None,
     write_out=False,
     output_base_path=None,
-    model_prompt=None
+    model_prompt=None,
+    stop_words=None,
+    apply_chat_template=False,
 ):
     """Instantiate and evaluate a model on a list of tasks.
 
@@ -255,7 +263,7 @@ def evaluate(
 
             ctx = MODEL_PROMPT_MAP[model_prompt](ctx)
             
-            reqs = task.construct_requests(doc, ctx)
+            reqs = task.construct_requests(doc, ctx, stop_words=stop_words, use_chat=apply_chat_template)
 
             if write_out:
                 prompt_details.append({"doc_id": doc_id})
diff --git a/src/tasks/__init__.py b/src/tasks/__init__.py
index e4bd6a1..44d384c 100644
--- a/src/tasks/__init__.py
+++ b/src/tasks/__init__.py
@@ -14,9 +14,11 @@ TASK_REGISTRY = {
     "flare_es_fns": flare.ESFNS,
     "flare_es_tsa": flare.ESTSA,
     "flare_fpb": flare.FPB,
+    "flare_fpb_inst": flare.FPBInst,
     "flare_fiqasa": flare.FIQASA,
     "flare_ner": flare.NER,
     "flare_finqa": flare.FinQA,
+    "flare_finqa_inst": flare.FinQAInst,
     "flare_convfinqa": flare.ConvFinQA,
     "flare_headlines": flare.Headlines,
     "flare_finer_ord": flare.FinerOrd,
@@ -32,6 +34,7 @@ TASK_REGISTRY = {
     "flare_multifin_en": flare.MultiFinEN,
     "flare_tsa": flare.TSA,
     "flare_cfa": flare.CFA,
+    "flare_cfa_inst": flare.CFAInst,
     "flare_ma": flare.MA,
     "flare_causal20_sc": flare.Causal20SC,
     "flare_finarg_ecc_arc": flare.FINARGECCARC,
diff --git a/src/tasks/flare.py b/src/tasks/flare.py
index 56cfb33..41e74f5 100644
--- a/src/tasks/flare.py
+++ b/src/tasks/flare.py
@@ -15,6 +15,8 @@ from factscore_package.factscorer import FactScorer
 import os
 #from comet import download_model, load_from_checkpoint
 
+from sympy import sympify
+
 _CITATION = """
 @misc{xie2023pixiu,
       title={PIXIU: A Large Language Model, Instruction Data and Evaluation Benchmark for Finance}, 
@@ -54,7 +56,7 @@ class Classification(Task):
     def test_docs(self):
         return self.dataset["test"]
 
-    def construct_requests(self, doc, ctx):
+    def construct_requests(self, doc, ctx, stop_words=None, use_chat=False):
         """Uses RequestFactory to construct Requests and returns an iterable of
         Requests which will be sent to the LM.
 
@@ -65,7 +67,7 @@ class Classification(Task):
             language description, as well as the few shot examples, and the question
             part of the document for `doc`.
         """
-        cont_request = rf.greedy_until(ctx, {"until": None})
+        cont_request = rf.greedy_until(ctx, {"until": stop_words, "chat": use_chat})
         return cont_request
 
     def doc_to_decontamination_query(self, doc):
@@ -203,7 +205,7 @@ class SequentialLabeling(Task):
             "entity_f1": True,
         }
 
-    def construct_requests(self, doc, ctx):
+    def construct_requests(self, doc, ctx, stop_words=None, use_chat=False):
         """Uses RequestFactory to construct Requests and returns an iterable of
         Requests which will be sent to the LM.
 
@@ -214,7 +216,7 @@ class SequentialLabeling(Task):
             language description, as well as the few shot examples, and the question
             part of the document for `doc`.
         """
-        cont_request = rf.greedy_until(ctx, {"until": None})
+        cont_request = rf.greedy_until(ctx, {"until": stop_words, "chat": use_chat})
         return cont_request
 
     def process_result(self, pred, gold, tokens):
@@ -319,7 +321,7 @@ class AbstractiveSummarization(Task):
             "bart_score": True,
         }
 
-    def construct_requests(self, doc, ctx):
+    def construct_requests(self, doc, ctx, stop_words=None, use_chat=False):
         """Uses RequestFactory to construct Requests and returns an iterable of
         Requests which will be sent to the LM.
 
@@ -330,7 +332,7 @@ class AbstractiveSummarization(Task):
             language description, as well as the few shot examples, and the question
             part of the document for `doc`.
         """
-        cont_request = rf.greedy_until(ctx, {"until": None})
+        cont_request = rf.greedy_until(ctx, {"until": stop_words, "chat": use_chat})
         return cont_request
 
     def rouge_score(self, items):
@@ -440,7 +442,7 @@ class ExtractiveSummarization(Task):
             "bart_score": True,
         }
 
-    def construct_requests(self, doc, ctx):
+    def construct_requests(self, doc, ctx, stop_words=None, use_chat=False):
         """Uses RequestFactory to construct Requests and returns an iterable of
         Requests which will be sent to the LM.
 
@@ -451,7 +453,7 @@ class ExtractiveSummarization(Task):
             language description, as well as the few shot examples, and the question
             part of the document for `doc`.
         """
-        cont_request = rf.greedy_until(ctx, {"until": None})
+        cont_request = rf.greedy_until(ctx, {"until": stop_words, "chat": use_chat})
         return cont_request
 
     def get_sum(self, labels, texts):
@@ -579,7 +581,7 @@ class RelationExtraction(Task):
             "f1": True,
         }
 
-    def construct_requests(self, doc, ctx):
+    def construct_requests(self, doc, ctx, stop_words=None, use_chat=False):
         """Uses RequestFactory to construct Requests and returns an iterable of
         Requests which will be sent to the LM.
 
@@ -590,7 +592,7 @@ class RelationExtraction(Task):
             language description, as well as the few shot examples, and the question
             part of the document for `doc`.
         """
-        cont_request = rf.greedy_until(ctx, {"until": None})
+        cont_request = rf.greedy_until(ctx, {"until": stop_words, "chat": use_chat})
         return cont_request
 
     def process(self, items):
@@ -669,7 +671,7 @@ class QA(Task):
         # TODO: Format the query prompt portion of the document example.
         return doc["query"]
 
-    def construct_requests(self, doc, ctx):
+    def construct_requests(self, doc, ctx, stop_words=None, use_chat=False):
         """Uses RequestFactory to construct Requests and returns an iterable of
         Requests which will be sent to the LM.
 
@@ -680,7 +682,7 @@ class QA(Task):
             language description, as well as the few shot examples, and the question
             part of the document for `doc`.
         """
-        cont_request = rf.greedy_until(ctx, {"until": None})
+        cont_request = rf.greedy_until(ctx, {"until": stop_words, "chat": use_chat})
         return cont_request
 
     def doc_to_target(self, doc):
@@ -710,6 +712,15 @@ class FPB(Classification):
     DATASET_PATH = "TheFinAI/flare-fpb"
 
 
+class FPBInst(FPB):
+    def doc_to_text(self, doc):
+        instruction_suffix = "Please answer with only one of the following labels: 'positive', 'negative', or 'neutral'. Do not output anything else."
+        last_characters = doc["query"][-7:]
+        if last_characters != "Answer:":
+            raise ValueError(f"Query must end with 'Answer:', but got `{last_characters}`.")
+        return doc["query"][:-7] + instruction_suffix
+
+
 class FIQASA(Classification):
     DATASET_PATH = "TheFinAI/flare-fiqasa"
 
@@ -751,7 +762,7 @@ class NER(Task):
         # TODO: Format the query prompt portion of the document example.
         return doc["query"]
 
-    def construct_requests(self, doc, ctx):
+    def construct_requests(self, doc, ctx, stop_words=None, use_chat=False):
         """Uses RequestFactory to construct Requests and returns an iterable of
         Requests which will be sent to the LM.
 
@@ -762,7 +773,7 @@ class NER(Task):
             language description, as well as the few shot examples, and the question
             part of the document for `doc`.
         """
-        cont_request = rf.greedy_until(ctx, {"until": None})
+        cont_request = rf.greedy_until(ctx, {"until": stop_words, "chat": use_chat})
         return cont_request
 
     def doc_to_target(self, doc):
@@ -794,6 +805,49 @@ class NER(Task):
 class FinQA(QA):
     DATASET_PATH = "TheFinAI/flare-finqa"
 
+    def process_results(self, doc, results):
+        gold = doc["answer"]
+        pred = results[0].strip()
+
+        acc = 1.0 if pred == gold else 0.0
+
+        # __within_eps function is cited from the implementation in FinanceReasoning.
+        # "FinanceReasoning: Benchmarking Financial Numerical Reasoning More Credible, Comprehensive and Challenging". Zichen Tang et al. ACL2025.Z
+        def __within_eps(pred: float, gt: float):
+            eps = abs(gt) * 0.002
+            if pred >= gt - eps and pred <= gt + eps:
+                return True
+            else:
+                return False
+
+        def __parse_string_to_float(string: str):
+            return float(sympify(string, evaluate=True).evalf())
+
+        try:
+            gold_val = __parse_string_to_float(gold)
+            pred_val = __parse_string_to_float(pred)
+            acc_within_eps = 1.0 if __within_eps(pred_val, gold_val) else 0.0
+        except Exception:
+            acc_within_eps = 0.0
+
+        return {
+            "acc": acc,
+            "acc_within_eps": acc_within_eps,
+        }
+
+    def aggregation(self):
+        return {
+            "acc": mean,
+            "acc_within_eps": mean,
+        }
+
+class FinQAInst(FinQA):
+    def doc_to_text(self, doc):
+        instruction_suffix = "Please answer with only the calculated value as a number. Do not output anything else."
+        last_characters = doc["query"][-7:]
+        if last_characters != "Answer:":
+            raise ValueError(f"Query must end with 'Answer:', but got `{last_characters}`.")
+        return doc["query"][:-7] + instruction_suffix
 
 class StockMovement(Classification):
     DATASET_NAME = None
@@ -888,7 +942,7 @@ class Headlines(Classification):
             all_f1s.append(f1)
         return np.mean(all_f1s)
 
-    def construct_requests(self, doc, ctx):
+    def construct_requests(self, doc, ctx, stop_words=None, use_chat=False):
         """Uses RequestFactory to construct Requests and returns an iterable of
         Requests which will be sent to the LM.
 
@@ -899,7 +953,7 @@ class Headlines(Classification):
             language description, as well as the few shot examples, and the question
             part of the document for `doc`.
         """
-        cont_request = rf.greedy_until(ctx, {"until": None})
+        cont_request = rf.greedy_until(ctx, {"until": stop_words, "chat": use_chat})
         return cont_request
 
     def aggregation(self):
@@ -1021,7 +1075,7 @@ class TSA(Task):
             "rmse": False,
         }
 
-    def construct_requests(self, doc, ctx):
+    def construct_requests(self, doc, ctx, stop_words="Answer:"):
         """
         Uses RequestFactory to construct Requests and returns an iterable of
         Requests which will be sent to the LM.
@@ -1033,7 +1087,7 @@ class TSA(Task):
             language description, as well as the few shot examples, and the question
             part of the document for `doc`.
         """
-        cont_request = rf.greedy_until(ctx, {"until": "Answer:"})
+        cont_request = rf.greedy_until(ctx, {"until": stop_words, "chat": use_chat})
         return cont_request
 
     def rmse(self, items):
@@ -1057,9 +1111,10 @@ class TSA(Task):
 
 
 
-class CFA(Classification):
-    DATASET_PATH = "TheFinAI/flare-cfa"
-    LOWER_CASE = False
+# Duplicated
+# class CFA(Classification):
+#     DATASET_PATH = "TheFinAI/flare-cfa"
+#     LOWER_CASE = False
 
 
 class FINARGECCARC(Classification):
@@ -1085,6 +1140,23 @@ class FSRL(SequentialLabeling):
 class CFA(Classification):
     DATASET_PATH = "TheFinAI/flare-cfa"
 
+    def has_training_docs(self):
+        return False
+
+    def has_validation_docs(self):
+        return False
+
+    def has_test_docs(self):
+        return True
+
+class CFAInst(CFA):
+    def doc_to_text(self, doc):
+        instruction_suffix = "Please answer with only a single letter: 'A', 'B', or 'C'. Do not output anything else."
+        last_characters = doc["query"][-7:]
+        if last_characters != "Answer:":
+            raise ValueError(f"Query must end with 'Answer:', but got `{last_characters}`.")
+        return doc["query"][:-7] + instruction_suffix
+
 class FinargECCAUC(Classification):
     DATASET_PATH = "TheFinAI/flare-finarg-ecc-auc"
 
@@ -1195,7 +1267,7 @@ class LongFormFactuality(Task):
             "factscore": True
         }
 
-    def construct_requests(self, doc, ctx):
+    def construct_requests(self, doc, ctx, stop_words=None, use_chat=False):
         """Uses RequestFactory to construct Requests and returns an iterable of
         Requests which will be sent to the LM.
         :param doc:
@@ -1205,7 +1277,7 @@ class LongFormFactuality(Task):
             language description, as well as the few shot examples, and the question
             part of the document for `doc`.
         """
-        cont_request = rf.greedy_until(ctx, {"until": None})
+        cont_request = rf.greedy_until(ctx, {"until": stop_words, "chat": use_chat})
         return cont_request
 
     def factscore(self, items):
